Sources/SwiftLexicalLookup/SequentialScopeSyntax.swift

```diff
@@ -0,0 +1,108 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+/// Scope that, in addition to names introduced by itself,
+/// also handles names introduced by
+/// `IntroducingToSequentialParentScopeSyntax` children scopes.
+protocol SequentialScopeSyntax: ScopeSyntax {}
+
+extension SequentialScopeSyntax {
+  /// Returns names introduced by `codeBlockItems`
+  /// and included `IntroducingToSequentialParentScopeSyntax` children
+  /// scopes that match the lookup.
+  ///
+  /// Example:
+  /// ```swift
+  /// func foo() {
+  ///   let a = 1
+  ///   guard let a = x else { return }
+  ///   let a = a // <-- 1
+  ///   guard let a = y else { return }
+  ///   a // <-- 2
+  /// }
+  /// ```
+  /// For the first `a` reference, sequential lookup returns
+  /// two results: from `guard` scope and from code block scope
+  /// in this exact order. For the second `a` reference,
+  /// sequential lookup yields results from four scopes starting
+  /// from the bottom: `guard`, code block, `guard` and
+  /// code block scope in this exact order.
+  func sequentialLookup(
+    in codeBlockItems: some Collection<CodeBlockItemSyntax>,
+    for identifier: Identifier?,
+    at origin: SyntaxProtocol,
+    with config: LookupConfig,
+    state: LookupState,
+    createResultsForThisScopeWith getResults: ([LookupName]) -> (LookupResult)
+  ) -> [LookupResult] {
+    var result: [LookupResult] = []
+    var currentChunk: [LookupName] = []
+    var itemsWithoutNamedDecl: [CodeBlockItemSyntax] = []
+
+    for codeBlockItem in codeBlockItems {
+      if Syntax(codeBlockItem.item).isProtocol(NamedDeclSyntax.self) {
+        currentChunk += LookupName.getNames(
+          from: codeBlockItem.item,
+          accessibleAfter: codeBlockItem.endPosition
+        ).filter { introducedName in
+          checkName(identifier, refersTo: introducedName, at: origin)
+        }
+      } else {
+        itemsWithoutNamedDecl.append(codeBlockItem)
+      }
+    }
+
+    for codeBlockItem in itemsWithoutNamedDecl {
```

Comment:
Would it make sense to do this in the `else` clause of the loop above? Also, technically we could have a type that conforms to `NamedDeclSyntax` and `IntroducingToSequentialParentScopeSyntax`.

Reply:
Sequential scope needs to ensure all type declarations are available in the whole scope (first loop) and then that results from `guard` (and possibly other `IntroducingToSequentialParentScopeSyntax` in the future) are properly interleaved with the results produced by this scope. That's also why we can't add names directly to `results` as then we would loose this sort of partitioning and proper interleaving with the results from `guard` etc. I think [this test](https://github.com/swiftlang/swift-syntax/blob/c83b3c6f3454a05b0a22ce26f12de594de19d68b/Tests/SwiftLexicalLookupTest/NameLookupTests.swift#L803-L832) nicely exemplifies this behavior.

Reply:
Hmm I think you have a point here. Though, I think it's a bit hard to nail all the nuances without comparing them directly to the current implementation. I think I'd prefer to wait with fixing details like this until we hook up `SwiftLexicalLookup` to the compiler and start testing it there.